From bc3b1314e514d87d1afd39ec5b8d4d6abecef834 Mon Sep 17 00:00:00 2001
Message-Id: <bc3b1314e514d87d1afd39ec5b8d4d6abecef834.1421390813.git.chang-joon.lee@intel.com>
In-Reply-To: <55165e324aa18c65d2b754de3f9d891b8a8477e9.1421390813.git.chang-joon.lee@intel.com>
References: <55165e324aa18c65d2b754de3f9d891b8a8477e9.1421390813.git.chang-joon.lee@intel.com>
From: Sagar Kamble <sagar.a.kamble@intel.com>
Date: Fri, 5 Dec 2014 18:06:42 +0530
Subject: [PATCH 4/6] REVERTME [VPG]: drm/i915/vlv: Additional rpm get/put to
 avoid suspend before first modeset

Due to an earlier patch, which reduced the auto-suspend delay to 500 ms,

Gfx device's runtime suspend was triggered before first modeset &
DPST init call from CoreU.
As the display was power gated on suspend, the Display registers were reset.
This was breaking DPST when it attempted to get the (now invalid) resolution
info from the registers. With this change Gfx won't suspend till the first
mode set call and once modeset happens, device won't suspend till next DPMS
off call.

The patch is marked as REVERTME, as this is an interim solution.
A more clean fix will be submitted later.

v2: Restricting the change to VLV.

Issue: GMINL-4595
Change-Id: I0139cb29a0ece4f9a8d6781cbf8f46b521062a15
Signed-off-by: Akash Goel <akash.goel@intel.com>
Signed-off-by: Sagar Kamble <sagar.a.kamble@intel.com>
---
 drivers/gpu/drm/i915/i915_dma.c      |    7 +++++++
 drivers/gpu/drm/i915/intel_display.c |   13 ++++++++++---
 2 files changed, 17 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_dma.c b/drivers/gpu/drm/i915/i915_dma.c
index 9ef6f89..167f9b1 100644
--- a/drivers/gpu/drm/i915/i915_dma.c
+++ b/drivers/gpu/drm/i915/i915_dma.c
@@ -1889,6 +1889,13 @@ int i915_driver_load(struct drm_device *dev, unsigned long flags)
 
 	intel_init_runtime_pm(dev_priv);
 
+	/*
+	 * Taking RPM reference here for device to stay awake till first
+	 * modeset happens. Will help initialize DPST properly.
+	*/
+	if (IS_VALLEYVIEW(dev) && !IS_CHERRYVIEW(dev))
+		intel_runtime_pm_get(dev_priv);
+
 	i915_perfmon_setup(dev_priv);
 
 	i915_perfmon_cleanup(dev_priv);
diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index f197b26..50eeaf0 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -11998,10 +11998,8 @@ static int __intel_set_mode(struct drm_crtc *crtc,
 
 	/* DO it only once */
 	if (IS_VALLEYVIEW(dev))
-		if (dev_priv->is_first_modeset) {
+		if (dev_priv->is_first_modeset)
 			program_pfi_credits(dev_priv, true);
-			dev_priv->is_first_modeset = false;
-		}
 
 	/* crtc->mode is already used by the ->mode_set callbacks, hence we need
 	 * to set it here already despite that we pass it down the callchain.
@@ -12029,6 +12027,15 @@ static int __intel_set_mode(struct drm_crtc *crtc,
 	if (dev_priv->display.modeset_global_resources)
 		dev_priv->display.modeset_global_resources(dev);
 
+	/* DO it only once */
+	if (IS_VALLEYVIEW(dev))
+		if (dev_priv->is_first_modeset) {
+			/* This will drop reference taken in i915_driver_load */
+			if (!IS_CHERRYVIEW(dev))
+				intel_runtime_pm_put(dev_priv);
+			dev_priv->is_first_modeset = false;
+		}
+
 	/* Set up the DPLL and any encoders state that needs to adjust or depend
 	 * on the DPLL.
 	 */
-- 
1.7.9.5

